astropy/timeseries/periodograms/lombscargle/implementations/cython_impl.pyx

```diff
@@ -74,15 +76,13 @@ def lombscargle_cython(t, y, dy, frequency, normalization='standard',
 
     # pre-center the data: not technically required if fit_mean=True,
     # but it simplifies the math.
+    w = dy ** -2
+    w /= np.sum(w)
```

Comment:
Sorry, but I'm not an expert in the tests, however, the main issue with the Cython method is that it's not compiled by default during the test execution, so it's not available when the tests are running.

I would be glad if anyone could suggest a potential solution to the issue without the main redesign of the testing unit, but for now, I'll probably leave it in its current form.

One idea I have (which isn't elegant but could work) would be to import the Cython method using the [pyximport](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiling-with-pyximport) during the testing phase, which would work, but again would add a lot of overhead to the tests unit.

Reply:
> doesn't this affect the result on line 83 ? (I realize this could be balanced out by another change downstream in `_lomb_scargle`, but I want to make sure it is)

And about that change - it does affect the result on line 83, but it's supposed to. Essentially it pre-normalizes weights, which allows for the removal of weight recalculation during every iteration of the main loop (what the current code does). So it offsets the lack of secondary normalization in `_lomb_scargle`, essentially leading to *almost* the same results, however, it marginally increases the numerical errors, as the weights in the previous iterations were recalculated.

It could've been fixed with Kahan's summation or other similar algorithm, but I guess it's not needed, as it still reliably falls within the accuracy bound defined in the tests unit.

Reply:
> And about that change - it does affect the result on line 83, but it's supposed to. Essentially it pre-normalizes weights, which allows for the removal of weight recalculation during every iteration of the main loop (what the current code does). So it offsets the lack of secondary normalization in _lomb_scargle, essentially leading to almost the same results, however, it marginally increases the numerical errors, as the weights in the previous iterations were recalculated.

Is it *needed* here or is it just nice-to-have™️ ? It sounds to me like this could be its own separate PR (and probably should)

about testing: it's a shame that the extension isn't available in CI builds, but it shouldn't stop us from writing a test now. Tests can be conditionally skipped if their requirements are missing using `@pytest.mark.skipif`. Don't hesitate to reach out if you need more guidance.

Reply:
> > And about that change - it does affect the result on line 83, but it's supposed to. Essentially it pre-normalizes weights, which allows for the removal of weight recalculation during every iteration of the main loop (what the current code does). So it offsets the lack of secondary normalization in _lomb_scargle, essentially leading to almost the same results, however, it marginally increases the numerical errors, as the weights in the previous iterations were recalculated.
> 
> Is it _needed_ here or is it just nice-to-have™️ ? It sounds to me like this could be its own separate PR (and probably should)
> 
> about testing: it's a shame that the extension isn't available in CI builds, but it shouldn't stop us from writing a test now. Tests can be conditionally skipped if their requirements are missing using `@pytest.mark.skipif`. Don't hesitate to reach out if you need more guidance.

Thanks for your reply. It's just a nice-to-have™️ but an industry standard implemented in all Astropy's Lomb-Scargle implementations except the Cython. For example in the 'slow' implementation lines 53 and 54 look like that: 
`w = dy**-2.0
w /= w.sum()`
    
While rewriting the code I've tried to keep the style relatively consistent with the rest of the implementations, as Cython clearly stood out, but I could also split it if it's required. Now it looks essentially almost like the 'fast' implementation, however with all of the approximations removed to keep the precision required from all of the non-fast methods.

As for testing the biggest issue is that the tests for Cython is already implemented. It's essentially tested in one of the iterations of the loop iterating over all available methods;
`ALL_METHODS_NO_AUTO = [method for method in ALL_METHODS if method != "auto"]`

The issue with that part of the code is that the Cython is only appended after compilation (and currently it's not compiled for the tests).

The approach I wanted to implement was to 'forcefully' append the Cython to the methods even if it's not compiled by adding a code block like that;
`    import pyximport
    pyximport.install(reload_support=True, language_level=3, setup_args={"include_dirs": [np.get_include()]})
    from ..implementations import cython_impl as cython_tmp
    # Append the Cython method to available methods if import succeeds
    ALL_METHODS = LombScargle.available_methods + ["cython_tmp"]`
    
 However, even that method seemingly failed, as even pyximport was not available during the tests. So I think leaving the tests how they are (by which I mean appending cython to the list only if it's compiled) looks like the best option, at least for now. If you have any idea how could it be improved I'll gladly accept any suggestions btw.

Reply:
> currently it's not compiled for the tests

actually I double checked that and I'm not sure what you mean here. It looks to me that `"cython"` is unconditionally included in `ALL_METHODS_NO_AUTO` (only `"auto"` is excluded, as well as `"scipy"` *if* it's missing). We *do* compile all extensions in CI, and that's also the normal thing to do if you're running tests locally. Are you not able to run these tests locally for some reason ?

Reply:
Sorry for the misunderstanding then. The issue I have both locally and in the repo is the line 193 of 'all optional dependecies' tests, which returns `Cython: not available` and doesn't compile the Cython implementation from the source. If I try to state it explicitly, at least locally the test fails, as Cython is not found.

But that may be either me misunderstanding the tests module (and the `Cython: not available` in full optional dependencies test run) or some weird kind of bug (and I have no clue what may be causing it in that case).

I mean I may try temporarily changing the Cython result to something wrong (for example to set the result to be equal to 1 for every frequency) and check if it'll pass tests in repo.

At least locally it does with no issues, even if Cython returns clearly wrong result. + that line 193 seems to explain that behavior, at least if it's not a local bug of my repo copy.

Reply:
Thanks, then I have no idea why locally the test always passes, even with wrong results, I could've messed up my local environment in some reason then (?).

Shall I try to run tests on the Cython version intentionally returning wrong results to see, if it'll fail (like it should've) or pass like my local copy?

Reply:
when editing cython code, you need to re-compile astropy for your change to take effect, otherwise tests will just import the latest compiled version of the extension module. Hope this helps !